streams/test-utils/src/test/java/org/apache/kafka/streams/test/TopologyTestDriverTest.java

```diff
@@ -0,0 +1,360 @@
+                sourceTopicName,
+                null,
+                null,
+                sourceTopicName,
+                sourceTopicName + "-processor",
+                new MockProcessorSupplier()
+            );
+        }
+
+        return topology;
+    }
+
+    @Test
+    public void shouldInitProcessor() {
+        testDriver = new TopologyTestDriver(setupSingleProcessorTopology(), config);
+        assertTrue(mockProcessors.get(0).initialized);
+    }
+
+    @Test
+    public void shouldCloseProcessor() {
+        testDriver = new TopologyTestDriver(setupSingleProcessorTopology(), config);
+
+        testDriver.close();
+        assertTrue(mockProcessors.get(0).closed);
+    }
+
+    @Test
+    public void shouldThrowForUnknownTopic() {
+        final String unknownTopic = "unknownTopic";
+        final ConsumerRecordFactory<byte[], byte[]> consumerRecordFactory = new ConsumerRecordFactory<>(
+            "unknownTopic",
+            new ByteArraySerializer(),
+            new ByteArraySerializer());
+
+        testDriver = new TopologyTestDriver(new Topology(), config);
+        try {
+            testDriver.process(consumerRecordFactory.create((byte[]) null));
+            fail("Should have throw IllegalArgumentException");
+        } catch (final IllegalArgumentException exception) {
+            assertEquals("Unknown topic: " + unknownTopic, exception.getMessage());
+        }
+    }
+
+    @Test
+    public void shouldProcessRecordForTopic() {
+        testDriver = new TopologyTestDriver(setupSourceSinkTopology(), config);
+
+        testDriver.process(consumerRecord1);
+        final ProducerRecord outputRecord = testDriver.readOutput(SINK_TOPIC);
+
+        assertEquals(key1, outputRecord.key());
+        assertEquals(value1, outputRecord.value());
+        assertEquals(SINK_TOPIC, outputRecord.topic());
+    }
+
+    @Test
+    public void shouldSetRecordMetadata() {
+        testDriver = new TopologyTestDriver(setupSingleProcessorTopology(), config);
+
+        testDriver.process(consumerRecord1);
+
+        final List<Record> processedRecords = mockProcessors.get(0).processedRecords;
+        assertEquals(1, processedRecords.size());
+
+        final Record record = processedRecords.get(0);
+        assertEquals(timestamp1, record.timestamp);
+        assertEquals(0, record.offset);
+        assertEquals(SOURCE_TOPIC_1, record.topic);
+    }
+
+    @Test
+    public void shouldSendRecordViaCorrectSourceTopic() {
+        testDriver = new TopologyTestDriver(setupMultipleSourceTopology(SOURCE_TOPIC_1, SOURCE_TOPIC_2), config);
+
+        final List<Record> processedRecords1 = mockProcessors.get(0).processedRecords;
+        final List<Record> processedRecords2 = mockProcessors.get(1).processedRecords;
+
+        testDriver.process(consumerRecord1);
+
+        assertEquals(1, processedRecords1.size());
+        assertEquals(0, processedRecords2.size());
+
+        Record record = processedRecords1.get(0);
```

Comment:
if you added an `equals(...)` to `Record` you could use `assertThat(expected, equalTo(..))`